File: hook/identity_test.go
package main

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// Same fixture table as packages/core/src/identity.test.ts -- both must
// canonicalize every one of these to the same string, or projectId
// diverges across languages the same way it diverged across SSH/HTTPS
// clones in production (2026-08-11).
var equivalentRemoteForms = []string{
	"git@github.com:Org/Repo.git",
	"https://github.com/Org/Repo.git",
	"https://github.com/Org/Repo",
	"https://github.com/Org/Repo/",
	"ssh://git@github.com/Org/Repo.git",
	"http://github.com/Org/Repo.git",
}

func TestCanonicalizeRemoteURL_EquivalentForms(t *testing.T) {
	want := "github.com/org/repo"
	for _, form := range equivalentRemoteForms {
		got := canonicalizeRemoteURL(form)
		if got != want {
			t.Errorf("canonicalizeRemoteURL(%q) = %q, want %q", form, got, want)
		}
	}
}

func TestCanonicalizeRemoteURL_DifferentReposStayDifferent(t *testing.T) {
	a := canonicalizeRemoteURL("git@github.com:Org/Repo.git")
	b := canonicalizeRemoteURL("git@github.com:Org/OtherRepo.git")
	if a == b {
		t.Errorf("expected different repos to canonicalize differently, both got %q", a)
	}
}

func TestCanonicalizeRemoteURL_SelfHostedSSHWithSubgroup(t *testing.T) {
	got := canonicalizeRemoteURL("git@gitlab.example.com:group/sub/repo.git")
	want := "gitlab.example.com/group/sub/repo"
	if got != want {
		t.Errorf("canonicalizeRemoteURL(...) = %q, want %q", got, want)
	}
}

// §17 design linking (2026-08): generateGroupID is a fresh, unpersisted
// primitive (unlike readOrCreatePersistedID above) -- just confirms it
// produces non-empty, distinct values on successive calls.
func TestGenerateGroupID_ReturnsNonEmptyDistinctValues(t *testing.T) {
	a := generateGroupID()
	b := generateGroupID()
	if a == "" || b == "" {
		t.Fatalf("generateGroupID() returned empty string(s): %q, %q", a, b)
	}
	if a == b {
		t.Errorf("two calls returned the same id %q, want distinct", a)
	}
}

// The phantom-.git guard, mirroring identity.test.ts's assertions that no
// `.git` is conjured outside a repo and nothing is written into a stray one.
// This port carried the bug those cover until 2026-09-18: readOrCreatePersistedID
// called os.MkdirAll, so a project id minted outside any repository left
// behind a directory that every existsSync(".git") check -- findRepoRoot's
// included -- reads as a repository forever after.

func TestReadOrCreatePersistedID_NeverCreatesTheDirectory(t *testing.T) {
	dir := t.TempDir()
	idPath := filepath.Join(dir, ".git", "twing-project-id")

	id := readOrCreatePersistedID(idPath)

	if id == "" {
		t.Error("an id is still returned for this run, just not persisted")
	}
	if _, err := os.Stat(filepath.Join(dir, ".git")); !os.IsNotExist(err) {
		t.Errorf("conjured a .git directory at %s; it must write into one, never create one", dir)
	}
}

func TestReadOrCreatePersistedID_WritesIntoAnExistingDirectory(t *testing.T) {
	dir := t.TempDir()
	if err := os.Mkdir(filepath.Join(dir, ".git"), 0o755); err != nil {
		t.Fatal(err)
	}
	idPath := filepath.Join(dir, ".git", "twing-project-id")

	first := readOrCreatePersistedID(idPath)
	second := readOrCreatePersistedID(idPath)

	if first != second {
		t.Errorf("id is not stable across calls: %q then %q", first, second)
	}
	data, err := os.ReadFile(idPath)
	if err != nil {
		t.Fatalf("expected the id to be persisted: %v", err)
	}
	if strings.TrimSpace(string(data)) != first {
		t.Errorf("persisted %q, returned %q", strings.TrimSpace(string(data)), first)
	}
}

// `.git` is a *file* in a worktree or submodule checkout. Writing a child
// path into it fails; the id must degrade to ephemeral rather than error.
func TestReadOrCreatePersistedID_ToleratesGitBeingAFile(t *testing.T) {
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, ".git"), []byte("gitdir: /elsewhere\n"), 0o644); err != nil {
		t.Fatal(err)
	}

	id := readOrCreatePersistedID(filepath.Join(dir, ".git", "twing-project-id"))

	if id == "" {
		t.Error("expected an ephemeral id rather than an empty one")
	}
}

func TestComputeProjectID_NoIDForSomewhereThatIsNotARepo(t *testing.T) {
	dir := t.TempDir()

	if got := computeProjectID(dir); got != "" {
		t.Errorf("minted %q for a directory that is no repository; a perfectly-shaped id for a project that has never existed is what sent the TS side querying a phantom for a month", got)
	}
	if _, err := os.Stat(filepath.Join(dir, ".git")); !os.IsNotExist(err) {
		t.Error("and it must leave no .git behind on the way out")
	}
}

// The half the TS side cannot hit: it takes an already-resolved repo root,
// this takes cwd. A repo with no origin, entered from a subdirectory, used to
// persist its id at <subdir>/.git/twing-project-id -- a phantom nested inside
// a real repo, so the same checkout answered with two different projects
// depending on where the hook happened to fire.
func TestComputeProjectID_NoRemoteFromSubdirUsesTheRepoRoot(t *testing.T) {
	root := t.TempDir()
	if _, ok := gitOutput(root, "init"); !ok {
		t.Skip("git unavailable")
	}
	sub := filepath.Join(root, "packages", "deep")
	if err := os.MkdirAll(sub, 0o755); err != nil {
		t.Fatal(err)
	}

	fromRoot := computeProjectID(root)
	fromSub := computeProjectID(sub)

	if fromRoot == "" || fromSub == "" {
		t.Fatalf("expected a persisted id for a real repo with no origin: %q, %q", fromRoot, fromSub)
	}
	if fromRoot != fromSub {
		t.Errorf("one checkout reported two projects: %q from the root, %q from %s", fromRoot, fromSub, sub)
	}
	if _, err := os.Stat(filepath.Join(sub, ".git")); !os.IsNotExist(err) {
		t.Errorf("conjured a nested .git at %s", sub)
	}
}

File: packages/core/src/identity.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { execFileSync, spawnSync } from "node:child_process";
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { canonicalizeRemoteUrl, parseGithubOwnerRepo, getOriginRemoteUrl, computeProjectId } from "./identity.js";

// Fixture table shared conceptually with hook/identity_test.go -- both must
// canonicalize every one of these to the same string, or projectId diverges
// across languages the same way it diverged across SSH/HTTPS clones in
// production (2026-08-11).
const EQUIVALENT_FORMS = [
  "git@github.com:Org/Repo.git",
  "https://github.com/Org/Repo.git",
  "https://github.com/Org/Repo",
  "https://github.com/Org/Repo/",
  "ssh://git@github.com/Org/Repo.git",
  "http://github.com/Org/Repo.git",
];

test("canonicalizeRemoteUrl: all equivalent clone forms produce the same result", () => {
  const results = EQUIVALENT_FORMS.map(canonicalizeRemoteUrl);
  const [first, ...rest] = results;
  for (const r of rest) {
    assert.equal(r, first);
  }
  assert.equal(first, "github.com/org/repo");
});

test("canonicalizeRemoteUrl: different repos stay different", () => {
  assert.notEqual(canonicalizeRemoteUrl("git@github.com:Org/Repo.git"), canonicalizeRemoteUrl("git@github.com:Org/OtherRepo.git"));
});

test("canonicalizeRemoteUrl: self-hosted git over ssh with a custom port-like path still normalizes the scp form", () => {
  assert.equal(canonicalizeRemoteUrl("git@gitlab.example.com:group/sub/repo.git"), "gitlab.example.com/group/sub/repo");
});

// §17 Phase 3
test("parseGithubOwnerRepo: extracts owner/repo from a canonicalized GitHub URL", () => {
  assert.deepEqual(parseGithubOwnerRepo(canonicalizeRemoteUrl("git@github.com:Org/Repo.git")), { owner: "org", repo: "repo" });
  assert.deepEqual(parseGithubOwnerRepo("github.com/twing-dev/twing-cli"), { owner: "twing-dev", repo: "twing-cli" });
});

test("parseGithubOwnerRepo: undefined for a non-GitHub host", () => {
  assert.equal(parseGithubOwnerRepo(canonicalizeRemoteUrl("git@gitlab.example.com:group/repo.git")), undefined);
});

test("parseGithubOwnerRepo: undefined for a malformed/incomplete GitHub path", () => {
  assert.equal(parseGithubOwnerRepo("github.com/just-an-org"), undefined);
  assert.equal(parseGithubOwnerRepo("github.com/org/repo/extra"), undefined);
});

// A repo with no `origin` remote is a normal, handled case (§17 Phase 3's
// A directory with no `.git` at all reaches the no-remote branch the same
// way a real repo without an origin does -- findRepoRoot hands back its own
// argument when the walk finds nothing -- and used to be given a freshly
// minted random id, which callers then queried the coordinator with. Found
// live 2026-09-16: `design list --server <url>` one directory above a repo
// answered "no designs", exit 0, a different invented id every call.
test("computeProjectId: refuses to invent an id for a directory that is not a repo", () => {
  const notARepo = fs.mkdtempSync(path.join(os.tmpdir(), "twing-identity-not-a-repo-"));
  try {
    assert.throws(() => computeProjectId(notARepo), /not a git repository[\s\S]*-C <path-to-repo>/);
    assert.equal(fs.existsSync(path.join(notARepo, ".git", "twing-project-id")), false, "and writes nothing");
  } finally {
    fs.rmSync(notARepo, { recursive: true, force: true });
  }
});

// How the phantom repo got made, and why it survived: a twing command run
// outside any repo resolved a root of $HOME, the no-remote fallback created
// $HOME/.git to hold its id, and every findRepoRoot walk from anywhere under
// $HOME stopped there from then on -- a directory that is not a repository
// but satisfies every existsSync(".git") check. Found on this developer's
// machine 2026-09-16, dated 2026-08-17.
test("computeProjectId: never conjures a .git directory, and rejects a stray one", () => {
  const notARepo = fs.mkdtempSync(path.join(os.tmpdir(), "twing-identity-phantom-"));
  try {
    assert.throws(() => computeProjectId(notARepo), /not a git repository/);
    assert.equal(fs.existsSync(path.join(notARepo, ".git")), false, "must not create .git");

    // And the shape that actually bit: a .git that exists but is not a repo.
    fs.mkdirSync(path.join(notARepo, ".git"));
    assert.throws(() => computeProjectId(notARepo), /not a git repository/, "a stray .git is not a repo, whatever existsSync says");
    assert.equal(fs.existsSync(path.join(notARepo, ".git", "twing-project-id")), false, "and nothing is written into it");
  } finally {
    fs.rmSync(notARepo, { recursive: true, force: true });
  }
});

test("computeProjectId: a real repo with no origin still gets its persisted id", () => {
  // The fallback this guard must not break: no remote means no way to clone,
  // so a gitignored random id per repo is correct (§8).
  const repoRoot = fs.mkdtempSync(path.join(os.tmpdir(), "twing-identity-no-remote-"));
  try {
    execFileSync("git", ["init", "--quiet"], { cwd: repoRoot });
    const first = computeProjectId(repoRoot);
    assert.match(first, /^[0-9a-f-]{36}$/);
    assert.equal(computeProjectId(repoRoot), first, "and it is stable across calls");
  } finally {
    fs.rmSync(repoRoot, { recursive: true, force: true });
  }
});

// no-remote fallback) -- git's own "fatal/error: No such remote 'origin'"
// isn't a real error here and shouldn't leak anywhere just because we
// happen to shell out to git to find that out (found live, 2026-08-18).
// This has to run getOriginRemoteUrl in a *child* process and inspect
// *that* child's stderr fd -- `execFileSync`'s default stdio duplicates the
// grandchild git process's stderr straight onto our fd 2, bypassing
// `process.stderr.write()`/anything else at the JS layer entirely, so
// nothing short of a real subprocess boundary can observe (or fail to
// observe) the leak this test exists to catch.
test("getOriginRemoteUrl: returns null and leaks nothing onto stderr for a repo with no origin remote", () => {
  const repoRoot = fs.mkdtempSync(path.join(os.tmpdir(), "twing-identity-test-"));
  try {
    execFileSync("git", ["init", "--quiet"], { cwd: repoRoot });
    const distDir = path.dirname(new URL(import.meta.url).pathname); // this test file's own compiled location, dist/ (tests run against dist, not src)
    const modulePath = path.join(distDir, "identity.js");
    const script = `
      import(${JSON.stringify(modulePath)}).then(({ getOriginRemoteUrl }) => {
        process.stdout.write(JSON.stringify(getOriginRemoteUrl(${JSON.stringify(repoRoot)})));
      });
    `;
    const child = spawnSync(process.execPath, ["-e", script], { encoding: "utf8" });
    assert.equal(child.stderr, "");
    assert.equal(child.stdout.trim(), "null");
  } finally {
    fs.rmSync(repoRoot, { recursive: true, force: true });
  }
});
